### ISSUE TITLE
Empty ecommerce items in datalayer for category and product pages when 'Use Flat Catalog' is enabled

### ISSUE THREAD

**Issue by amjadm61** [yireo/Yireo_GoogleTagManager2#160]
When I activate 'Use Flat Catalog Category' in the store->configuration->Catalog->Catalog settings, the items in the e-commerce datalayer are empty for the category and product pages. However, when I change the 'Use Flat Catalog' setting to 'NO', everything works as expected.

**Comment by jissereitsma** [comment]
Which version of this module and which version of Magento is this?

**Comment by amjadm61** [comment]
Extension version: 3.3.1
Magento version: Adobe Commerce 2.4.5-p2

The problem is occurring due to the catch method in the file https://github.com/yireo/Yireo_GoogleTagManager2/blob/master//DataLayer/Mapper/ProductDataMapper.php#L107
try {
$categories = $this->getCategoryFromProduct->getAll($product);
} catch (NoSuchEntityException $e) {
return [];
}
In this catch block, the array is being cleared with an empty array. Instead, it should return the $data.

**Comment by jissereitsma** [comment]
First of all, I think this is a pretty big bug, which I'll fix right away: https://github.com/yireo/Yireo_GoogleTagManager2/commit/0db53440dae7c135b204a10fd9e78b9a0b53a0d4

Another point I want to make is that the flat catalog is deprecated since 2.3 (ref https://experienceleague.adobe.com/docs/commerce-admin/catalog/catalog/catalog-flat.html) so it would be better to keep it disabled.

However, I still fail to see where the involved code of `GetCategoryFromProduct::getAll()` is failing. I've tested this myself and in my case, it is simply working :( Under 2.4.6 Open Source though. I'm still busy with testing more ...

**Comment by jissereitsma** [comment]
I've done some further testing and can't find any reason why `GetCategoryFromProduct::getAll()` would fail under 2.4.x with the Flat Catalog enabled. The call itself does a simple repository fetching with 2 columns filtered (`entity_id` and `is_active`) and both fields are there both in the EAV tables and the flat tables ...

**Comment by jissereitsma** [comment]
Perhaps it is a third-party add on in your specific shop?

**Comment by amjadm61** [comment]
I was just trying out this extension to see how it's working and noticed this category issue. I didn't get time to look into this in detail. By the way, I found the root cause of the issue:

We have a multi-website setup with 2 root categories, X and Y. X is for the first website, and Y is for the second website. The product is assigned to the first website but has categories assigned from both root categories. So, the getAll() method was trying to access the category that belongs to the second root category Y as well.

We can fix this issue by fetching the rootCategoryId from the current store and using it in the filter builder, like this:

```
    public function getAll(ProductInterface $product): array
    {
        $categoryIds = $product->getCategoryIds();
        $rootCategoryId = $this->storeManager->getStore()->getRootCategoryId();

        $this->searchCriteriaBuilder = $this->searchCriteriaBuilderFactory->create();

        $entityIdFilterGroup = $this->filterGroupBuilder->create();
        $entityIdFilterGroup->setFilters([$this->filterBuilder
            ->setField('entity_id')
            ->setConditionType('in')
            ->setValue($categoryIds)
            ->create()]);

        $rootCategIdFilterGroup = $this->filterGroupBuilder->create();
        $rootCategIdFilterGroup->setFilters([$this->filterBuilder
            ->setField('parent_id')
            ->setConditionType('eq')
            ->setValue($rootCategoryId)
            ->create()]);

        $filterGroup = $this->filterGroupBuilder->create();
        $filterGroup->setFilters([$this->filterBuilder
            ->setField('is_active')
            ->setConditionType('eq')
            ->setValue(1)
            ->create()]);

        $this->searchCriteriaBuilder->setFilterGroups([$entityIdFilterGroup, $rootCategIdFilterGroup, $filterGroup]);
        $searchCriteria = $this->searchCriteriaBuilder->create();

        return $this->categoryListRepository->getList($searchCriteria)->getItems();
    }
```

If you want, I can create a PR for this with the fix.

**Comment by jissereitsma** [comment]
I believe that fix is not the right one. The check is there for a subcategory to have the root category as a **direct** parent, while there could be other parents in between. Forget about the PR, since I'm working on it already. I think the simplest would be to check for the right website scope.

**Comment by jissereitsma** [comment]
Still slightly different - a match for the `path` - but same-same-but-different: https://github.com/yireo/Yireo_GoogleTagManager2/commit/c26f585905e0f940d65a5f6c80f88ea7f22e8e99

**Comment by jissereitsma** [comment]
I'm closing this issue because it should be fixed in 3.3.2+. Let me know if it is not fixed yet :)
